cameo/flux_analysis/structural.py

```diff
@@ -16,25 +16,118 @@
 
 from __future__ import print_function
 
+import logging
 from copy import copy
 
-import sympy
-from cameo.core import SolverBasedModel
-from cameo import Reaction
-from cobra import Metabolite
+import numpy as np
 import optlang
 import six
-import logging
-
-logger = logging.getLogger(__name__)
-logger.setLevel(logging.DEBUG)
+import sympy
+from cobra import Metabolite
+from numpy.linalg import svd
 
+from cameo import Reaction
+from cameo.core import SolverBasedModel
 from cameo.exceptions import SolveError, Infeasible
 from cameo.util import TimeMachine
 
+from scipy.cluster.hierarchy import linkage
+
+
 __all__ = ['find_dead_end_reactions', 'find_coupled_reactions', 'ShortestElementaryFluxModes']
 
 
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+# Taken from http://wiki.scipy.org/Cookbook/RankNullspace
+def nullspace(matrix, atol=1e-13, rtol=0):
+    """Compute an approximate basis for the nullspace of A.
+
+    The algorithm used by this function is based on the singular value
+    decomposition of `A`.
+
+    Parameters
+    ----------
+    matrix : ndarray
+        A should be at most 2-D.  A 1-D array with length k will be treated
+        as a 2-D with shape (1, k)
+    atol : float
+        The absolute tolerance for a zero singular value.  Singular values
+        smaller than `atol` are considered to be zero.
+    rtol : float
+        The relative tolerance.  Singular values less than rtol*smax are
+        considered to be zero, where smax is the largest singular value.
+
+    If both `atol` and `rtol` are positive, the combined tolerance is the
+    maximum of the two; that is::
+        tol = max(atol, rtol * smax)
+    Singular values smaller than `tol` are considered to be zero.
+
+    Return value
+    ------------
+    ns : ndarray
+        If `A` is an array with shape (m, k), then `ns` will be an array
+        with shape (k, n), where n is the estimated dimension of the
+        nullspace of `A`.  The columns of `ns` are a basis for the
+        nullspace; each element in numpy.dot(A, ns) will be approximately
+        zero.
+    """
+
+    matrix = np.atleast_2d(matrix)
+    u, s, vh = svd(matrix)
+    tol = max(atol, rtol * s[0])
+    nnz = (s >= tol).sum()
+    ns = vh[nnz:].conj().T
+    return ns
+
+
+def find_blocked_reactions_nullspace(model, ns=None, tol=1e-10):
+    """Identify reactions that can't carry flux based on the nullspace of the stoichiometric matrix.
+    A blocked reaction will correspond to an all-zero row in N(S)."""
+    if ns is None:
+        ns = nullspace(model.S)
+    mask = (np.abs(ns) <= tol).all(1)
+    blocked = [reac for reac, b in zip(model.reactions, mask) if b is True]
```

Comment:
If you use the `future` package. You can `from builtins import zip` which will give you a version of `zip` in Python2 & 3 that works like `itertools.izip`, i.e., does not create a list first.